File: app/MMarket/c/src/pgs/simulationsPg.c
// Copyright 29-Jan-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "pgs/simulationsPg.h"
#include "dmc/cgi.h"
#include "dmc/str.h"
#include "dmc/err.h"
#include "dmc/js.h"
#include "data/models.h"
#include "db/simProfitsDb.h"

char *simulationsPg_process (Mchar *mrq) {
  char *rq = cgi_rq_string(mrq, "rq");
  if (str_eq(rq, "idata")) {
    char *model_id = cgi_rq_string(mrq, "modelId");

    Model *md = models_get(model_id);
    ASimProfitsRow *profits = simProfitsDb_read(model_id)->rows;
    Mchar *rp = mchar_new();
    mchar_put(rp, "model", model_to_js(md));
    mchar_put(rp, "profits", aSimProfitsRow_to_js(profits));
    return cgi_rp(rp);
  } else return FAIL(str_f("Unexpected value for 'rq': %s", rq));
}

File: app/KtMMarket/go/pgs/simulationsPg/simulationsPg.go
// Copyright 04-Jul-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

// Simulations page.
package simulationsPg

import (
	"github.com/dedeme/KtMMarket/data/model"
	"github.com/dedeme/KtMMarket/db"
	"github.com/dedeme/ktlib/cgi"
	"github.com/dedeme/ktlib/js"
)

func Process(ck string, mrq cgi.T) string {
	rq := js.Rs(mrq["rq"])
	switch rq {
	case "idata":
		modelId := js.Rs(mrq["modelId"])

		md := model.FromId(modelId)
		profits := js.Ra(db.SimProfitsDb(modelId).ReadJs())[1]
		return cgi.Rp(ck, cgi.T{
			"model":   model.ToJs(md),
			"profits": profits,
		})
	default:
		panic("Value of rq (" + rq + ") is not valid")
	}
}